src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -0,0 +1,195 @@
+     * @return boolean useAPI or not.
+     */
+    private boolean setSizeFromAPI(String repoUrl, Item context, String credentialsId) {
+        List<RepositorySizeAPI> acceptedRepository = Objects.requireNonNull(RepositorySizeAPI.all())
+                .stream()
+                .filter(r -> r.isApplicableTo(repoUrl, context, credentialsId))
+                .collect(Collectors.toList());
+
+        if (acceptedRepository.size() > 0) {
+            try {
+                for (RepositorySizeAPI repo: acceptedRepository) {
+                    long size = repo.getSizeOfRepository(repoUrl, context, credentialsId);
+                    if (size != 0) { sizeOfRepo = size; }
+                }
+            } catch (Exception e) {
+                LOGGER.log(Level.INFO, "Not using performance improvement from REST API: " + e.getMessage());
+                return false;
+            }
+            return sizeOfRepo != 0; // Check if the size of the repository is zero
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Recommend a git implementation on the basis of the given size of a repository
+     * @param sizeOfRepo: Size of a repository (in KiBs)
+     * @return a git implementation, "git" or "jgit"
+     */
+    private String determineSwitchOnSize(Long sizeOfRepo, String gitExe) {
+        if (sizeOfRepo != 0L) {
+            if (sizeOfRepo >= SIZE_TO_SWITCH) {
+                return determineToolName(gitExe, "git");
+            } else {
+                return determineToolName(gitExe, JGitTool.MAGIC_EXENAME);
+            }
+        }
+        return "NONE";
+    }
+
+    /**
+     * For a given recommended git implementation, validate if the installation exists and provide no suggestion if
+     * implementation doesn't exist.
+     * @param gitImplementation: The recommended git implementation, "git" or "jgit" on the basis of the heuristics.
+     */
+    private void determineGitTool(String gitImplementation, String gitExe) {
+        if (gitImplementation.equals("NONE")) {
+            gitTool = "NONE";
+            return; // Recommend nothing (GitToolRecommendation = NONE)
+        }
+        final Jenkins jenkins = Jenkins.get();
+        GitTool tool = GitUtils.resolveGitTool(gitImplementation, jenkins, null, TaskListener.NULL);
+        if (tool != null) {
+            if (!gitExe.equals(tool.getGitExe()) && !tool.getGitExe().contains(JGitTool.MAGIC_EXENAME)) {
+                // for cases where git home is expanded, like: /usr/bin/git
+                gitTool = gitExe;
+                return;
+            }
+            gitTool = tool.getGitExe();
+        }
+    }
+
+    /**
+     * Recommend git tool to be used by the git client
+     * @return git implementation recommendation in the form of a string
+     */
+    public String getGitTool() {
+        return gitTool;
+    }
+
+    private String getCacheEntry(String remote) {
+        return "git-" + Util.getDigestOf(remote);
```

Comment:
Where does this logic come from? Is that something just for this?

Reply:
`getCacheEntry` is a static method in `AbstractGitSCMSource`, I have removed it from this class and borrowed it directly.